zookeeper/source/zk_nn_client.cc

```diff
@@ -881,47 +1058,76 @@ namespace zkclient{
+
+		for (auto repl : to_replicate) {
+			std::string read_from;
+			std::vector<std::string> target_dn;
+			std::uint64_t block_id;
+			std::stringstream strm(repl);
+			strm >> block_id;
+			if (!find_datanode_with_block(repl, read_from, err)) {
+				LOG(ERROR) << CLASS_NAME << " Failed to find datanode with this block! " << repl << " is lost!";
+				return false;
+			}
+			uint64_t blocksize;
+			if (!get_block_size(block_id, blocksize)) {
+				LOG(ERROR) << CLASS_NAME << "Replicate could not read the block size for block: " << block_id;
+				return false;
+			}
+			if (!find_datanode_for_block(target_dn, block_id, 1, false, blocksize) || target_dn.size() == 0) {
+				LOG(ERROR) << CLASS_NAME << " Failed to find datanode for this block! " << repl;
+				return false;
+			}
+			auto queue = REPLICATE_QUEUES + target_dn[0];
+			auto repl_item = util::concat_path(queue, repl);
+			ops.push_back(zk->build_create_op(repl_item, ZKWrapper::EMPTY_VECTOR));
+			auto read_from_item = util::concat_path(repl_item, read_from);
+			ops.push_back(zk->build_create_op(read_from_item, ZKWrapper::EMPTY_VECTOR));
+		}
 
-        // Create a payload vector containing the block_uuid
-        std::vector<std::uint8_t> data;
-        data.resize(sizeof(u_int64_t));
-        memcpy(&data[0], &block_uuid, sizeof(u_int64_t));
-
-        for (auto dn_id : datanodes) {
-            // The path of the sequential node, ends in "-"
-            std::string replicate_block_path = WORK_QUEUES + REPLICATE_BACKSLASH + dn_id + "/block-";
-
-            // Create an item on the replica queue for this block replica on the given datanode
-            std::string new_path;
-            if (!zk->create_sequential(replicate_block_path,
-                                       data,
-                                       new_path,
-                                       false,
-                                       error_code)) {
-                LOG(ERROR) << CLASS_NAME << "Failed to create_seq: " << replicate_block_path;
-                return false;
-            }
-            LOG(INFO) << CLASS_NAME << "Created: " << new_path;
+		if (!zk->execute_multi(ops, results, err)){
+			LOG(ERROR) << "Failed to execute multiop for replicate_blocks";
+			return false;
+		}
+	}
+
+
+    bool ZkNnClient::find_datanode_with_block(const std::string &block_uuid_str, std::string &datanode, int &error_code) {
+        std::vector<std::string> datanodes;
+        std::string block_loc_path = BLOCK_LOCATIONS + block_uuid_str;
+
+        if (!zk->get_children(block_loc_path, datanodes, error_code)) {
+            LOG(ERROR) << "Failed to get children of: " << block_loc_path;
+            return false;
+        }
+        if (datanodes.size() < 1) {
+            LOG(ERROR) << "There are no datanodes with a replica of block " << block_uuid_str;
+            // TODO: set error_code
+            return false;
         }
+        // TODO: Pick the datanode which has fewer transmits
```

Comment:
@prb2 @kyler-m we should have a set that contains the datanodes with the min number of xmits, and then randomly select one from that set